packages/discount-code-importer/src/main.js

```diff
@@ -1,3 +1,221 @@
+import { createUserAgentMiddleware }
+  from '@commercetools/sdk-middleware-user-agent'
+import { createSyncDiscountCodes } from '@commercetools/sync-actions'
+import { version } from '../package.json'
+
+export default class DiscountCodeImport {
+  // Set flowtype annotations
+  batchSize: number;
+  continueOnProblems: boolean;
+  client: Client;
+  apiConfig: AuthMiddlewareOptions;
+  logger: LoggerOptions;
+  _summary: Summary;
+  syncDiscountCodes: SyncAction;
+
+  constructor (
+    logger: LoggerOptions,
+    options: ConstructorOptions,
+  ) {
+    this.batchSize = options.batchSize || 50
+    this.apiConfig = options.apiConfig
+    this.continueOnProblems = options.continueOnProblems || false
+    this.client = createClient({
+      middlewares: [
+        createAuthMiddlewareForClientCredentialsFlow(this.apiConfig),
+        createUserAgentMiddleware({
+          libraryName: 'discount-code-importer',
+          libraryVersion: version,
+        }),
+        createHttpMiddleware({ host: this.apiConfig.host }),
+      ],
+    })
+
+    this.syncDiscountCodes = createSyncDiscountCodes()
+
+    this.logger = logger || {
+      error: npmlog.error.bind(this, ''),
+      warn: npmlog.warn.bind(this, ''),
+      info: npmlog.info.bind(this, ''),
+      verbose: npmlog.verbose.bind(this, ''),
+    }
+    this._resetSummary()
+  }
+
+  // Use static method because this is not called on any instance
+  static _buildPredicate (codeObjects: CodeDataArray): string {
+    const predicateArray = codeObjects.map(codeObject => codeObject.code)
+    return `code in ("${predicateArray.join('", "')}")`
+  }
+
+  performStream (chunk: ChunkOptions, cb: () => mixed) {
+    this._processBatches(chunk)
+      .then(() => cb())
+      .catch(err => cb(err.body))
+  }
+
+  _processBatches (codes: CodeDataArray) {
+    // Batch to `batchSize` to reduce necessary fetch API calls
+    const batchedList = _.chunk(codes, this.batchSize)
+    return Promise.map(batchedList, (codeObjects: CodeDataArray) => {
+      // Build predicate and fetch existing code
+      const predicate = DiscountCodeImport._buildPredicate(codeObjects)
+      const service = createRequestBuilder({
+        projectKey: this.apiConfig.projectKey,
+      })
+      const uri = service.discountCodes
+        .where(predicate)
+        .perPage(this.batchSize)
+        .build()
+      return this.client.execute({
+        uri,
+        method: 'GET',
+      })
+        .then((response: Object) => {
+          const existingCodes = response.body.results
+          return this._createOrUpdate(codeObjects, existingCodes)
+        })
+    }, { concurrency: 1 }) // Run one batch at a time
+      .then(() => Promise.resolve(this.summaryReport()))
+      .catch(error => Promise.reject({
+        error,
+        summary: this._summary,
+      }),
+      )
+  }
+
+  _createOrUpdate (newCodes: CodeDataArray, existingCodes: CodeDataArray) {
+    return Promise.map(newCodes, (newCode: CodeData) => {
```

Comment:
Since we already added concurrency above, we already limit the number of simultaneous request that the module can send to the API so I am not sure concurrency makes much sense here